lab/lab-guide.md

```diff
@@ -0,0 +1,802 @@
+
+This workflow for RAG Pattern generally maps to the following steps:
+
+1. User types in a user prompt or question.
+1. The user prompt is vectorized by an embeddings modeland returned as an array of vectors.
+1. These vectors are used in a vector search against adatabase. Results are returned ordered by semanticsimilarity.
+1. The search results, the context window (chathistory) and the latest user prompt are sent to the LLM.
+1. The LLM processes all of the text in the payload andgenerates a response.
+
+## Vector Search on user data
+
+The first step is to implement the vector search query in our application.
+
+1. Navigate to the **SemanticKernelService.cs** class and locate the **SearchProductsAsync()** function with the following signature. 
+
+    ```csharp
+    public async Task<string> SearchProductsAsync(ReadOnlyMemory<float> promptVectors, int productMaxResults)
+    ```
+
+1. Instead of writing a custom vector query to execute against our container, we can leverage Semantic Kernel's Azure Cosmos DB NoSQL Vector Store connector. This greatly reduces the amount of code we need to write in order to search our product data for relevant results. Comment out the two lines of code below, and add the following lines.
+
+    ```csharp
+    //string productsString = "";
+    //await Task.Delay(0);
+
+    var options = new VectorSearchOptions { VectorPropertyName = "vectors", Top = productMaxResults };
+    #pragma warning disable SKEXP0020
+    var searchResult = await _productContainer.VectorizedSearchAsync(promptVectors, options);
+    #pragma warning restore SKEXP0020
+    ```
+
+    This code does a few things including:
+
+    1. Uses `Top` in the `VectorSearchOptions` to limit the number of products that are returned by the search. Because LLM's can only process so much text at once, it is necessary to limit the amount of data returned by a vector search. The `productMaxResults` value limits that amount of data and because this is something you will need to adjust when doing vector searches, it is config value in this application.
+    1. Uses the `VectorizedSearchAsync()` function in the Semantic Kernel connector to perform vector search using the passed-in vectors generated by the user prompt. This automatically orders the results by the similarity score by most semantically relevant to least relevant.
+    1. Some functionality in Semantic Kernel is still experimental. We are disabling warnings for these experimental features to use them in our application. As the API surface area stabilizes in later package versions, this may no longer be required.
+
+1. After this block of code, add the following lines before the existing return statement. This loops through the vector search results and serializes all products as a single string.
+
+    ```csharp
+    var resultRecords = new List<VectorSearchResult<Product>>();
+    await foreach (var result in searchResult.Results)
+    {
+        resultRecords.Add(result);
```

Comment:
Can this be changed to AddRange() instead of putting through a for-each loop? I wasn't sure.

Reply:
This doesn't work because `searchResults.Results` is an `IAsyncEnumerable`. We could do something like this, but it's bad practice to use blocking calls because it's less performant than `await` which you have now. I think this function is fine to leave as is

```
resultRecords.AddRange(searchResult.Results.ToBlockingEnumerable());
```